Make wording and tone more consistent (#193)
Shopify/ruby-style-guide: README.md

* Code in a functional way. Avoid mutation (side effects) when you can.

* [Avoid defensive programming](http://www.erlang.se/doc/programming_rules.shtml#HDR11)

> * Keep method arguments unchanged.

Reviewer: I think this could be phrased using `avoid`, as you've done in other places
```suggestion
* Avoid mutating arguments, unless that is the purpose of the method.
```

Author: Applied! Except the "unless that is the purpose of the method" part since we shouldn't be appending "except when it makes sense to break the rule" because that's implicit.

* Use spaces around the `=` operator when assigning default values to method
  parameters.

> * Avoid line continuation `\` where not required.

Author: Removed the last sentence since it doesn't add anything to the rule imo.

## Syntax

* Use `::` only to reference constants (this includes classes and modules) and

>   constructors (like `Array()` or `Nokogiri::HTML()`). Avoid `::` for

Author: It's not up to the style guide to make assumptions about "how strong this rule should be". I am using "avoid" as the consistent verb to advise developers about things they shouldn't do, and making exceptions depending on how strong we feel about a specific rule defeats the purpose of this very change.

>   constructors (like `Array()` or `Nokogiri::HTML()`). Avoid `::` for

Reviewer 2: Ah okay, then I misunderstood its purpose. In that case I think there are still some cases that are formulated in a way that suggests how strong this should be but this is good to ship without that imo

  ternary operators must not be nested. Prefer if/else constructs in these
  cases.

> * Avoid multiline `?:` (the ternary operator); use `if/unless` instead.

Author: I'm just moving this rule to be closer to the one regarding ternary operators.

* Use `when x then ...` for one-line cases.

* Use `!` instead of `not`.

* Prefer `&&`/`||` over `and`/`or`. [More info on `and/or` for control

>   flow](http://devblog.avdi.org/2014/08/26/how-to-use-rubys-english-andor-operators-without-going-nuts/).

Author: Removed the link as I find it unnecessary.

Reviewer 2: Not sure, I think it's interesting. Like I vaguely remember that there was some weird difference esp. folks new to Ruby might be like "but why?"

Reviewer: The [difference involves precedence](https://stackoverflow.com/questions/1426826/difference-between-and-and-in-ruby), but it's not too hard to search for. Still, it might make sense to still have a link.

> * Avoid multiline `?:` (the ternary operator); use `if/unless` instead.

Author: I'm just moving this rule to be closer to the one regarding ternary operators.

    * `raise`
    * `puts`

> * Use class methods instead of a rails scope with a multi-line lambda

Reviewer 2: Do we have a rails style guide somewhere where this is metnioned?

> * Use class methods instead of a rails scope with a multi-line lambda

Author: We don't have a Rails Style Guide yet.

  names.map(&:upcase)
  ~~~

* Prefer `{...}` over `do...end` for single-line blocks.

* Prefer `do..end` over `{...}` for multi-line blocks.

* Omit `return` where possible.

* Omit `self` where possible.

  ~~~ ruby
  # bad
  self.my_method

  # good
  my_method

  # also good
  attr_writer :name

  def my_method
    self.name = 'Rafael' # `self` is neeeded to reference the attribute writer.
  end
  ~~~

* Wrap assignment in parentheses when using its return value in a conditional
  statement.

  ~~~ ruby
  if (value = /foo/.match(string))
  ~~~

* Use `||=` to initialize variables only if they're not already initialized.

> * Avoid using `||=` to initialize boolean variables.

Reviewer: Might be worth extending this from just booleans to also values that could be `nil`.

Reviewer 2: it _should_ extend to `nil` as `||` triggers that way for both `false` and `nil` so that the computation would be triggered again

Reviewer: Yeah. Consider

```ruby
def value
  @value ||= sleep(1000) && ['string', true, false, nil].shuffle
end
```

The truthy cases will be memoized, but the falsey ones won't be and would duplicate the computation, whereas this would memoize properly the first time:

```ruby
def value
  return @value if defined?(@value)
  @value = sleep(1000) && [true, false, nil].shuffle
end
```

Really, the rule should be

> Do not use `||=` to **memoize** potentially falsey variables (`false` or `nil`).
> &lt;code example demonstrating use of defined?&gt;

If the use case is potentially falsey memoization, then you **never** want to use `||=`. If you're using `||=` for something else, it **might** be fine.

  @enabled = true unless defined?(@enabled)
  ~~~

> * Avoid spaces between a method name and the opening parenthesis.

Reviewer 2: I think this should also be kept as `do not`

Reviewer: You might be right, as it can sometimes be a syntax error.

```ruby
def foo (a, b) # works
end

foo 1, 2       # works
foo(1, 2)      # works
foo (1, 2)     # SyntaxError: unexpected ',', expecting ')'
               # foo (1, 2)
               #       ^
```

Author: From my previous reply:

> It's not up to the style guide to make assumptions about "how strong this rule should be". I am using "avoid" as the consistent verb to advise developers about things they shouldn't do, and making exceptions depending on how strong we feel about a specific rule defeats the purpose of this very change.

  end
  ~~~

> * Avoid hashes-as-optional-parameters in general. Does the method do too much?

Reviewer 2: Did this move somewhere that I didn't see or do we want to get rid of it?

Author: I removed it because it's redundant with the rule below.

* Prefer `map` over `collect`, `find` over `detect`, `select` over `find_all`,
  `size` over `length`.

> * Prefer `Time` over `DateTime`.

Reviewer 2: Imo saying TZ vs offsets provides a nice and short rational

Author: With this change I'm removing such bits of rationale from rules to make them more concise. The fact that `Time` supports time zones is not the reason why people should follow the rule. It's because it makes code more consistent (like all the other rules).

* When defining binary operators, name the parameter `other`(`<<` and `[]` are
  exceptions to the rule, since their semantics are different).

> * Name predicate methods with a question mark. Predicate methods are methods

Reviewer 3: I'd replace the `question mark` to  `?` . In other part of this document, characters such as slash, brackets etc... aren't written literally

>   boolean value.

Reviewer 2: Imo it's also an imortant property that they don't have a side effect but iirc that's somewhat contentious

Author: Although I very much agree with you I think mentioning that would mean introducing new rules to the style guide, which I am avoiding in this PR.

* Avoid ending method names with a `?` if they don't return a boolean.

* Avoid prefixing method names with `is_`.

  ~~~ruby
  # bad
  def is_empty?
  end

  # good
  def empty?
  end
  ~~~

> * Avoid magic numbers. Use a constant and give it a meaningful name.

Reviewer 2: 👍

* Avoid nomenclature that has (or could be interpreted to have) discriminatory
  origins.

## Comments

> * When writing comments account for the fact that readers might not have the

Reviewer: ```suggestion
* Include relevant context in comments, as readers might be missing it.
```

>   same context as you do.

Reviewer 2: I don't think it's up yet, but soon there will be a new section in the development handbook dealing with source code comments (the google doc is already completely reviewed)

Author: Good to know!

> * Keep surrounding comments in sync when refactoring code.

Reviewer: Arguably, refactoring is the one case where the comments might not need to change. 😅 

```suggestion
* Keep comments in sync with code.
```

> * Write good copy, and use proper capitalization and punctuation in comments.

Reviewer: Personally, I'm not sure what "good copy" means in this context.

```suggestion
* Write comments using proper capitalization and punctuation.
```

> * Avoid superfluous comments. Focus on **why** the code is the way it is if

Author: Merged these three rules into a simpler single version.

  end
  ~~~

* Respect the [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)

>   When designing class hierarchies.

Reviewer 3: ```suggestion
  when designing class hierarchies.
```

  end
  ~~~

* Prefer `attr_reader` and `attr_accessor` over `attr`.

> * Avoid class (`@@`) variables.

Reviewer 2: here I also like the little reference to the reason, esp. people for new to ruby it might be nice so to see it's not just arbitrary rules but there's actual danger in using these.

Author: Same as above. Consistency is the reason why you should follow this rule.

  hash = {}
  ~~~

* Prefer the literal array syntax to `%w`, except when it reads substantially

>   more clearly in context.

Author: "except when..." is implicit not just for the rule but for pretty much everything in this style guide.

  STATES = ['draft', 'open', 'closed']
  ~~~

> * Append a trailing comma in multi-line collection literals.

Reviewer 2: the change is pretty hard from "it is encouraged" to "do it" maybe we could add a prefer or something. Unless ofc we want that ALL the time.

Author: Yes, and that's exactly why I'm changing the wording from a passive voice into the imperative tense. We shouldn't give space for doubt, the style guide should be as straightforward as possible.

Reviewer 2: I realize the intention of the PR was different and that's my misunderstanding, sorry - carry on :D

  makes diffs smaller and more meaningful.
* Append a trailing comma in multi-line collection literals.

  ~~~ ruby
  # not encouraged
  # bad
  {
    foo: :bar,
    baz: :toto
  }

  # encouraged
  # good
  {
    foo: :bar,
    baz: :toto,
  }
  ~~~

> * Prefer the literal array syntax to `%i`.

Author: Merge this with the rule at the top.

* Use the Ruby 1.9 hash literal syntax when your hash keys are symbols.

  ~~~ ruby
  # bad
  { :a => 1, :b => 2 }

  # good
  { a: 1, b: 2 }
  ~~~

> * Avoid mixing the Ruby 1.9 hash syntax with hash rockets in the same hash

Reviewer: We could probably merge this with the previous rule by saying something like this and combining the examples.

```markdown
* Use shorthand hash literal syntax when **all** keys are symbols
```

I also don't think referring to Ruby 1.9 is very clear, as most developers at Shopify will have started writing Ruby well into 2.x.

  puts "$global = #{$global}"
  ~~~

> * Avoid `Object#to_s` on interpolated objects.

Reviewer 2: Imo that's another `don't` case because it's directly what it does, I can't imagine needing this in any scenario, but maybe it's my imagination

  /(?:first|second)/
  ~~~

> * Prefer `Regexp#match` over Perl-legacy variables to capture group matches.

Reviewer: Keeping

> (`$1`, `$2`, etc)

would make it clearer what this refers to for people who haven't used Perl.

Author: I used the code sample to exemplify `$1`. We should always prefer to clarify a rule with a code sample with necessary (instead of using extra sentences or parentheses).

Reviewer 2: Like that line of thinking!

* Treat test code like any other code you write. This means: keep readability,
  maintainability, complexity, etc. in mind.

> * Use Minitest as test framework.

Reviewer 2: It's missing `the` right? I wish this would be `prefer` but I understand 👼

* Limit each test case to cover a single aspect of your code.

> * Organize the setup, action, and assertion sectinos of the test case with empty

Reviewer 3: ```suggestion
* Organize the setup, action, and assertion sections of the test case with empty
```

> * Organize the setup, action, and assertion sections of the test case with empty

Reviewer: ```suggestion
* Organize the setup, action, and assertion sections of the test case into paragraphs separated by empty
```

* Avoid needless metaprogramming.

> * Avoid starting method names with `get_`.

Reviewer 2: feels like this could be moved up to where we discuss predicate/? methods

> * Avoid ending method names with `!` when there is no equivalent method without

Reviewer 2: also would probably move it up to the `?` section

  the bang. Bangs are to mark a more dangerous version of a method, e.g. `save`
  returns a `bool` in ActiveRecord, whereas `save!` will throw an exception on

>   failure.

Reviewer: Might be worth mentioning the mutation use case as well (e.g. `map` vs `map!`).

Also "`bool`" should probably just be "boolean"

Author: Replaced `bool` with "boolean". I'm not mentioning the mutation use case because that's the one we are encouraging people to avoid.
